File: Tasks/ford-fulkerson.cpp
#include <iostream>
#include <queue>
#include <climits>

const int V = 6;

bool bfs(int graph[V][V], int s, int t, int parent[]) {
    bool visited[V] = {false};
    std::queue <int> queue;
    queue.push(s);
    visited[s] = true;
    parent[s] = -1;

    while (!queue.empty()) {
        int u = queue.front();
        queue.pop();
        for (int v = 0; v < V; v++) {
            if (!visited[v] && graph[u][v] > 0) {
                queue.push(v);
                parent[v] = u;
                visited[v] = true;
            }
        }
    }
    return visited[t];
}

int fordFulkerson(int graph[V][V], int s, int t) {
    int u, v;
    int graph1[V][V];
    for (u = 0; u < V; u++) {
        for (v = 0; v < V; v++)
            graph1[u][v] = graph[u][v];
    }

    int parent[V];
    int max_flow = 0;

    while (bfs(graph1, s, t, parent)) {
        int path_flow = INT_MAX;
        for (v = t; v != s; v = parent[v]) {
            u = parent[v];
            path_flow = std::min(path_flow, graph1[u][v]);
        }
        for (v = t; v != s; v = parent[v]) {
            u = parent[v];
            graph1[u][v] -= path_flow;
            graph1[v][u] += path_flow;
        }
        max_flow += path_flow;
    }
    
    return max_flow;
}

int main() {
     int graph[V][V] = { 
        {0, 11, 12, 0, 0, 0}, 
        {0, 0, 0, 12, 0, 0}, 
        {0, 1, 0, 0, 11, 0}, 
        {0, 0, 0, 0, 0, 19}, 
        {0, 0, 0, 7, 0, 4}, 
        {0, 0, 0, 0, 0, 0} 
        }; 
  
    std::cout << fordFulkerson(graph, 0, 5) << std::endl; 
  
    return 0; 
}

File: Tasks/ford-fulkerson.py
from collections import deque
 
class Graph: 
   
    def __init__(self,graph): 
        self.graph = graph 
        self.vertices_count = len(graph) 
   
    def bfs(self, root, target, parent): 
        visited = [False] * self.vertices_count 
        queue = deque() 
        queue.append(root) 
        visited[root] = True

        while queue: 
            u = queue.popleft() 
            for ind, val in enumerate(self.graph[u]): 
                if not visited[ind] and val > 0: 
                    queue.append(ind)
                    visited[ind] = True
                    parent[ind] = u

        return visited[target]
    

    def fordFulkerson(self, source, sink): 
        parent = [-1] * self.vertices_count
        max_flow = 0
        while self.bfs(source, sink, parent):
            path_flow = float("Inf")
            s = sink
            while(s != source):
                path_flow = min(path_flow, self.graph[parent[s]][s])
                s = parent[s]
            max_flow += path_flow
            v = sink
            while(v != source):
                u = parent[v]
                self.graph[u][v] -= path_flow
                self.graph[v][u] += path_flow
                v = parent[v]
        return max_flow 
  

graph_as_matrix = [
        [0, 11, 12, 0, 0, 0],
        [0, 0, 0, 12, 0, 0], 
        [0, 1, 0, 0, 11, 0], 
        [0, 0, 0, 0, 0, 19], 
        [0, 0, 0, 7, 0, 4], 
        [0, 0, 0, 0, 0, 0]] 
  
graph = Graph(graph_as_matrix) 
  
source = 0
sink = 5
   
print (graph.fordFulkerson(source, sink)) 
